### PR TITLE
Tickets/dm 27242

### PR DESCRIPTION
**Pull request by tribeiro**

### CODE REVIEW

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst:13]: I think it's worth adding a sentence here WHY a user would want to write a SAL script. Feel free to massage the suggestion below.
"Having the ability to execute the SAL Script via the ScriptQueue is recommended when the script will be launched either multiple times, in concert with other scripts, or to make it easily available and executable by other operators"

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst]: "may be interested" -- might want to be a little stronger here and use "recommended"  -- also might want to put in a plug for tstn-010?

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst]: *takes a series ..

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst]: This is basically the same content as in line 22. 
"As anticipated in.."  might be better written as "If your procedure follows what was anticipated in the ...
I'd suggest just re-organizing this as follows:

This procedure follows from a real life scenario where a user needed functionality that dithered the telescope in a specific pattern and performed a series of observations at each position. 
As anticipated in the `Concept of Operations`_,  the procedure was first developed using a Jupyter notebook, and tested it in one of the `Operational Environments`_ (including on sky tests).

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst]: being done there = being performed by the script ?

What aspects do you need them to understand?

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst:53]: Add a sentence first about why you chose external scripts.
e.g. Because this script is not going to be used as part of regular operations, and the functionality is still in development, it should be located in the `external_scripts`_ repository.

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst]: I think you do tickets "on" a branch not "in" as that'd imply being inside the wood of the branch. 🌳

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst]: Maybe make it "ticket branch" -- meaning put the quotes around ticket and branch?
I agree this terminology is pretty unclear...

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst]: the bit about IDE's and text editors are in the note below... probably just easiest to nix it here?
Option #2 is incorporate your note here.

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst]: I think VSCode is basically what everyone is moving to... 
https://developer.lsst.io/editors/vscode.html

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst]: *know

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst]: methods

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst:131]: Do you want to say that formatting here follows numpy docstrings?

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst]: what is this description for and what does it do?

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst:189]: random thought...
I wonder if we could create a script that would make a script... kind-of like sqr-bot junior...
might not be worth it....

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [reply Control-User-Interfaces/writing-sal-scripts.rst]: to work on that is consistent with the name of the JIRA ticket

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst:240]: and to use the python executable in /usr/bin/env - which is mandatory for our system

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst]: hmm so your call but I'd write this as
... statements in line 24 is derived from the location of your SAL script in the package where the leading ``python`` is removed and ``.`` replaces ``/`` (e.g. .....).

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst]: so this is not obvious to me why it's necessary. Is it worth explaining in a sentence?

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst]: Now THAT we have a base Python module, we are ready to start filling in some code.
.. or "filling it up with some code"

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst]: Let's break down the schema header line-by-line.

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst]: this case? all cases? this feels like magic...

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst]: Note that the path above will not actually exist until we merge the code back to the master branch.

Comment (patrickingraham) [reply Control-User-Interfaces/writing-sal-scripts.rst]: I'm confused.. master or develop?
I thought we did everything off the develop branch and it worked fine...

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst]: link to semantic versioning?

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst]: always use?

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst]: descriptions by placing the >- on the first line and indenting from the "description" attribute.

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst]: uhh silly question... where do the config files go? I've always launched from a notebook and never actually written a config file for a script!

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst:369]: So this is from the atcs or atptg, right?
Might be best to say that this field is from the atptg component and info can be found there. However, then say that you'll use this as an opportunity to show how someone can build a complex description?

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst]: I don't think this is true? 
It feels like this is probably in the wrong place ?

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst:632]: Is it worth adding that the scriptQueue will dump a script if it exceeds these estimates?
I think you want to add something about how accurate they need to be

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst:718]: what about using `arun` versus `run` such that the script doesn't close out all the remotes when it's finished?

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst:891]: I didn't know these existed!

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [inline Control-User-Interfaces/writing-sal-scripts.rst]: runs

File: Control-User-Interfaces/example-notebook/ExampleNotebook.rst
Comment (patrickingraham) [inline Control-User-Interfaces/example-notebook/ExampleNotebook.rst:4]: This doesn't appear to be rendering correctly... none of the code blocks are showing..

**Review by patrickingraham — CHANGES_REQUESTED** [review]
This looks really great!
Most comments are just minor stuff. The only major thing is that the auto-generated notebook doesn't appear to be rendering correctly. No code is shown.
Also, should we make a point of telling people they should practice running their scripts on the NTS before going on-sky?

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (tribeiro) [reply Control-User-Interfaces/writing-sal-scripts.rst:189]: That is an interesting idea... we could definitely look into that.

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (tribeiro) [reply Control-User-Interfaces/writing-sal-scripts.rst:240]: I think the directive above will not actually use the python executable in /usr/bin/env but find the correct python version that is setup in the environment at the time the script runs.

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (tribeiro) [reply Control-User-Interfaces/writing-sal-scripts.rst]: uhm.. you tell me... the explanation may be a bit beyond the scope here. Maybe we could improve the documentation in salobj about this and point the user there?

https://ts-salobj.lsst.io/sal_scripts.html#python-script-file

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (tribeiro) [reply Control-User-Interfaces/writing-sal-scripts.rst]: well, develop is the branch used for development... we do releases on the master branch.

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (tribeiro) [reply Control-User-Interfaces/writing-sal-scripts.rst]: not a silly question at all... So, as you can see from how we operate scripts from a notebook, SAL Scripts are configured "on the fly", by sending the configuration over DDS or directly calling methods that runs the configuration procedures. When we run from the ScriptQueue we pass the configuration as a yaml string to the ScriptQueue which later sends it to the SAL Script over DDS. 

So, basically, the configuration are logged to DDS (and therefore stored in the EFD), but we don't have a repo for Script configuration. I think this kind of make sense as some Scripts can be configured in any number of different ways.

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (tribeiro) [reply Control-User-Interfaces/writing-sal-scripts.rst:369]: This is actually from the `ATCS`.. I will link the documentation.

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (tribeiro) [reply Control-User-Interfaces/writing-sal-scripts.rst:632]: I don't think the ScriptQueue will do that. That may be a nice feature, though it would require some good thinking. 

We don't have a specific accuracy needed. At this point this is mainly used for information purposes.

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (tribeiro) [reply Control-User-Interfaces/writing-sal-scripts.rst:718]: I am not sure if we want to get into this level of detail here. It is not necessarily part of the SAL Script base interface. Maybe mention it in an appendix like section?

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [reply Control-User-Interfaces/writing-sal-scripts.rst]: so maybe skirt this and say "if the value is not encountered in the user-provided configuration" ?
the term "file" is a bit misleading...

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [reply Control-User-Interfaces/writing-sal-scripts.rst:632]: Maybe say, "it is helpful to have reasonable time estimates as they are displayed via LOVE and watched by observers" or something like that..

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [reply Control-User-Interfaces/writing-sal-scripts.rst:718]: ok, we can leave that for another day... but I think we want to make it a "default" essentially...

File: Control-User-Interfaces/writing-sal-scripts.rst
Comment (patrickingraham) [reply Control-User-Interfaces/writing-sal-scripts.rst]: discussed in private, is here by accident.

**Review by patrickingraham — APPROVED** [review]
Looks good. See a couple follow-up comments before merging.

**Pull request merged**
